File: p2p-ws-sdk-dart/lib/src/core_compat/core_frame.dart
import "dart:typed_data";

Uint8List encodeCoreFrame({required int magic, required Uint8List payload}) {
  final out = Uint8List(8 + payload.length);
  final bd = ByteData.sublistView(out);
  bd.setInt32(0, payload.length, Endian.big);
  bd.setInt32(4, magic, Endian.big);
  out.setRange(8, out.length, payload);
  return out;
}

({int magic, Uint8List payload}) decodeCoreFrame(Uint8List frameBytes) {
  if (frameBytes.length < 8) {
    throw FormatException("frame too short");
  }
  final bd = ByteData.sublistView(frameBytes);
  final len = bd.getInt32(0, Endian.big);
  final magic = bd.getInt32(4, Endian.big);
  final end = 8 + len;
  if (len < 0 || end > frameBytes.length) {
    throw FormatException("bad length");
  }
  return (magic: magic, payload: Uint8List.sublistView(frameBytes, 8, end));
}

List<({int magic, Uint8List payload})> decodeAllCoreFrames(Uint8List bytes) {
  final out = <({int magic, Uint8List payload})>[];
  var pos = 0;
  while (pos < bytes.length) {
    if (pos + 8 > bytes.length) {
      throw FormatException("frame truncated");
    }
    final bd = ByteData.sublistView(bytes, pos, pos + 8);
    final len = bd.getInt32(0, Endian.big);
    final magic = bd.getInt32(4, Endian.big);
    final end = pos + 8 + len;
    if (len < 0 || end > bytes.length) {
      throw FormatException("bad length");
    }
    out.add((magic: magic, payload: Uint8List.sublistView(bytes, pos + 8, end)));
    pos = end;
  }
  return out;
}

File: p2p-ws-sdk-ts/src/core_compat/core_frame.ts
import { Buffer } from "node:buffer"

export function encodeCoreFrame(magic: number, payload: Uint8Array): Buffer {
  const buf = Buffer.allocUnsafe(8 + payload.length)
  buf.writeInt32BE(payload.length, 0)
  buf.writeInt32BE(magic | 0, 4)
  Buffer.from(payload).copy(buf, 8)
  return buf
}

export function decodeCoreFrame(frame: Uint8Array): { magic: number; payload: Uint8Array } {
  if (frame.byteLength < 8) throw new Error("frame too short")
  const buf = Buffer.from(frame)
  const len = buf.readInt32BE(0)
  const magic = buf.readInt32BE(4)
  if (len < 0 || 8 + len > buf.length) throw new Error("bad length")
  return { magic, payload: buf.subarray(8, 8 + len) }
}

export function decodeAllCoreFrames(frame: Uint8Array): Array<{ magic: number; payload: Uint8Array }> {
  const buf = Buffer.from(frame)
  const out: Array<{ magic: number; payload: Uint8Array }> = []
  let pos = 0
  while (pos < buf.length) {
    if (pos + 8 > buf.length) throw new Error("frame truncated")
    const len = buf.readInt32BE(pos)
    const magic = buf.readInt32BE(pos + 4)
    const end = pos + 8 + len
    if (len < 0 || end > buf.length) throw new Error("bad length")
    out.push({ magic, payload: buf.subarray(pos + 8, end) })
    pos = end
  }
  return out
}
